### PR TITLE
LPD-26861 make sure that a relationship field will not have its information processed before it is treated properly

### PR DESCRIPTION
**Pull request by vngoliveira**
[LPD-26861](https://liferay.atlassian.net/browse/LPD-26861)

### CODE REVIEW

File: modules/apps/object/object-info-api/src/main/java/com/liferay/object/info/item/provider/util/ObjectEntryInfoItemValuesProviderUtil.java [hunk 1]
```diff
@@ -91,17 +91,17 @@ public static List<InfoFieldValue<Object>> getInfoFieldValues(
...
 
+				_addInfoFieldValue(
+					dlAppLocalService, dlURLHelper, infoFieldValues,
+					listTypeEntryLocalService, objectEntryLocalService,
+					objectField, objectFieldInfoFieldConverter,
+					ObjectField.class.getSimpleName(),
+					objectRelationshipLocalService, themeDisplay,
+					values.get(objectField.getName()));
+
 				continue;
```
Comment (carolmariaabb) [inline modules/apps/object/object-info-api/src/main/java/com/liferay/object/info/item/provider/util/ObjectEntryInfoItemValuesProviderUtil.java:105]: @vngoliveira we should include the info field value for relationship fields. Please try to fix the code where the NPE is happening. 

In the [SiteObjectScopeProviderImpl](https://github.com/liferay/liferay-portal/blob/bdae4717b6a049554992063ab13a1fb1035d1bac/modules/apps/object/object-service/src/main/java/com/liferay/object/internal/scope/SiteObjectScopeProviderImpl.java#L35), you can try to do something like:

```
@Override
public long getGroupId(HttpServletRequest httpServletRequest)
	throws PortalException {

	ThemeDisplay themeDisplay =
		(ThemeDisplay)httpServletRequest.getAttribute(
			WebKeys.THEME_DISPLAY);

	if (themeDisplay != null) {
		return themeDisplay.getScopeGroupId();
	}

	ServiceContext serviceContext =
		ServiceContextThreadLocal.getServiceContext();

	if (serviceContext == null) {
		return 0;
	}

	return serviceContext.getScopeGroupId();
}
```

File: modules/apps/object/object-info-api/src/main/java/com/liferay/object/info/item/provider/util/ObjectEntryInfoItemValuesProviderUtil.java
Comment (carolmariaabb) [inline modules/apps/object/object-info-api/src/main/java/com/liferay/object/info/item/provider/util/ObjectEntryInfoItemValuesProviderUtil.java:1]: @vngoliveira please send the PR with an integration test

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java [hunk 2]
```diff
@@ -114,12 +114,8 @@ public static void tearDownClass() {
...
 		_group = GroupTestUtil.addGroup();
 
-		ServiceContext serviceContext =
-			ServiceContextTestUtil.getServiceContext(_group.getGroupId());
-
-		serviceContext.setRequest(_getHttpServletRequest(_group));
-
-		ServiceContextThreadLocal.pushServiceContext(serviceContext);
+		_serviceContext = ServiceContextTestUtil.getServiceContext(
+			_group.getGroupId());
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:118]: @vngoliveira this variable is only used by one test: `testObjectEntryInfoItemFieldValuesProvider`.

That said, I see no reason to make it global and add it here in the `setUp` method.

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java [hunk 3]
```diff
@@ -131,6 +127,8 @@ public void setUp() throws Exception {
 				Collections.singletonList(
 					ListTypeEntryUtil.createListTypeEntry(_listTypeEntryKey)));
 
+		_mockHttpServletRequest = new MockHttpServletRequest();
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:130]: @vngoliveira this variable is only used by one test `testObjectEntryInfoItemFieldValuesProvider`.

That said, I see no reason to make it global and add it here in the `setUp` method.

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java [hunk 4]
```diff
@@ -114,12 +114,8 @@ public static void tearDownClass() {
 	public void setUp() throws Exception {
 		_group = GroupTestUtil.addGroup();
 
-		ServiceContext serviceContext =
-			ServiceContextTestUtil.getServiceContext(_group.getGroupId());
-
-		serviceContext.setRequest(_getHttpServletRequest(_group));
-
-		ServiceContextThreadLocal.pushServiceContext(serviceContext);
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:122]: @vngoliveira we still have the `ServiceContextThreadLocal.popServiceContext()` in the `tearDown` method. But you removed the `ServiceContextThreadLocal.pushServiceContext(serviceContext)` from the  `setUp` method.

That said, we also need remove the `ServiceContextThreadLocal.popServiceContext()` from the `tearDown` method and change the test to something like:

```
ServiceContextThreadLocal.pushServiceContext(serviceContext);

try {
   // all the logic
}
finally {
   ServiceContextThreadLocal.popServiceContext();
}

```

Comment (carolmariaabb) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:122]: This way we guarantee that only what was "pushed" will be removed from the stack.

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java [hunk 5]
```diff
@@ -241,9 +239,64 @@ public void testObjectEntryInfoItemFieldValuesProvider() throws Exception {
...
+		_assertValues(
+			fileEntry, infoItemFieldValues1, objectAction,
+			parentTextObjectFieldNameValue);
+
+		// Getting infoItemFieldValues passing the HttpServletRequest with a
+		// ThemeDisplay attribute
+
+		_serviceContext.setRequest(_getHttpServletRequest(_group));
+
+		ServiceContextThreadLocal.pushServiceContext(_serviceContext);
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:261]: @vngoliveira notice that you are calling the "push" method twice here. So you also need to call the "pop" twice.

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java [hunk 6]
```diff
@@ -241,9 +239,64 @@ public void testObjectEntryInfoItemFieldValuesProvider() throws Exception {
 			).build(),
 			false);
 
-		InfoItemFieldValues infoItemFieldValues =
+		// Getting infoItemFieldValues passing the HttpServletRequest without a
+		// ThemeDisplay attribute
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:243]: @vngoliveira simplify to `HttpServletRequest with null themeDisplay`

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java [hunk 7]
```diff
@@ -241,9 +239,64 @@ public void testObjectEntryInfoItemFieldValuesProvider() throws Exception {
...
+
+		InfoItemFieldValues infoItemFieldValues1 =
 			infoItemFieldValuesProvider.getInfoItemFieldValues(objectEntry);
 
+		_assertValues(
+			fileEntry, infoItemFieldValues1, objectAction,
+			parentTextObjectFieldNameValue);
+
+		// Getting infoItemFieldValues passing the HttpServletRequest with a
+		// ThemeDisplay attribute
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:257]: @vngoliveira simplify to `HttpServletRequest with themeDisplay`

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java [hunk 8]
```diff
@@ -241,9 +239,64 @@ public void testObjectEntryInfoItemFieldValuesProvider() throws Exception {
...
+		ObjectField objectField, Object expectedValue) {
+
+		InfoFieldValue<Object> fileNameInfoFieldValue =
+			infoItemFieldValues.getInfoFieldValue(
+				objectField.getObjectFieldId() + infoFieldName);
+
+		Assert.assertEquals(expectedValue, fileNameInfoFieldValue.getValue());
+	}
+
+	private void _assertValues(
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:295]: @vngoliveira rename it to `_assertInfoFieldValues`

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java [hunk 9]
```diff
@@ -241,9 +239,64 @@ public void testObjectEntryInfoItemFieldValuesProvider() throws Exception {
...
+		_serviceContext.setRequest(_mockHttpServletRequest);
+
+		ServiceContextThreadLocal.pushServiceContext(_serviceContext);
+
+		InfoItemFieldValues infoItemFieldValues1 =
 			infoItemFieldValuesProvider.getInfoItemFieldValues(objectEntry);
 
+		_assertValues(
+			fileEntry, infoItemFieldValues1, objectAction,
+			parentTextObjectFieldNameValue);
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:254]: @vngoliveira inline:

```
_assertInfoFieldValues(
	fileEntry, infoItemFieldValuesProvider.getInfoItemFieldValues(objectEntry),
	objectAction, parentTextObjectFieldNameValue);
```

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java [hunk 10]
```diff
@@ -241,9 +239,64 @@ public void testObjectEntryInfoItemFieldValuesProvider() throws Exception {
...
+		_serviceContext.setRequest(_getHttpServletRequest(_group));
+
+		ServiceContextThreadLocal.pushServiceContext(_serviceContext);
+
+		InfoItemFieldValues infoItemFieldValues2 =
+			infoItemFieldValuesProvider.getInfoItemFieldValues(objectEntry);
+
+		_assertValues(
+			fileEntry, infoItemFieldValues2, objectAction,
+			parentTextObjectFieldNameValue);
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:268]: @vngoliveira inline:

```
_assertInfoFieldValues(
	fileEntry, infoItemFieldValuesProvider.getInfoItemFieldValues(objectEntry),
	objectAction, parentTextObjectFieldNameValue);
```

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:1]: @vngoliveira my suggestion:

```
@Test
public void testObjectEntryInfoItemFieldValuesProvider() throws Exception {
	FileEntry fileEntry = _dlAppLocalService.addFileEntry(
		null, TestPropsValues.getUserId(), _group.getGroupId(),
		DLFolderConstants.DEFAULT_PARENT_FOLDER_ID, "test.txt",
		ContentTypes.TEXT, RandomTestUtil.randomBytes(), null, null, null,
		ServiceContextTestUtil.getServiceContext(_group.getGroupId()));

	String parentTextObjectFieldNameValue = RandomTestUtil.randomString();

	ObjectEntry parentObjectEntry = _objectEntryLocalService.addObjectEntry(
		TestPropsValues.getUserId(), _group.getGroupId(),
		_parentObjectDefinition.getObjectDefinitionId(),
		HashMapBuilder.<String, Serializable>put(
			"parentTextObjectFieldName", parentTextObjectFieldNameValue
		).build(),
		ServiceContextTestUtil.getServiceContext());

	ObjectEntry objectEntry = _objectEntryLocalService.addObjectEntry(
		TestPropsValues.getUserId(), _group.getGroupId(),
		_childObjectDefinition.getObjectDefinitionId(),
		HashMapBuilder.<String, Serializable>put(
			"r_oneToManyRelationshipName_" +
				_parentObjectDefinition.getPKObjectFieldName(),
			parentObjectEntry.getObjectEntryId()
		).put(
			"attachmentObjectFieldName", fileEntry.getFileEntryId()
		).put(
			"picklistObjectFieldName", _listTypeEntryKey
		).build(),
		ServiceContextTestUtil.getServiceContext());

	ObjectAction objectAction = _objectActionLocalService.addObjectAction(
		RandomTestUtil.randomString(), TestPropsValues.getUserId(),
		objectEntry.getObjectDefinitionId(), true, StringPool.BLANK,
		RandomTestUtil.randomString(),
		LocalizedMapUtil.getLocalizedMap(RandomTestUtil.randomString()),
		LocalizedMapUtil.getLocalizedMap(RandomTestUtil.randomString()),
		RandomTestUtil.randomString(),
		ObjectActionExecutorConstants.KEY_WEBHOOK,
		ObjectActionTriggerConstants.KEY_STANDALONE,
		UnicodePropertiesBuilder.put(
			"url", RandomTestUtil.randomString()
		).build(),
		false);

	// HttpServletRequest with null themeDisplay

	_testObjectEntryInfoItemFieldValuesProvider(
		fileEntry, objectAction, objectEntry,
		parentTextObjectFieldNameValue, null);

	// HttpServletRequest with themeDisplay

	ThemeDisplay themeDisplay = new ThemeDisplay();

	themeDisplay.setCompany(
		_companyLocalService.getCompany(_group.getCompanyId()));
	themeDisplay.setLocale(LocaleUtil.getDefault());
	themeDisplay.setScopeGroupId(_group.getGroupId());
	themeDisplay.setSiteGroupId(_group.getGroupId());
	themeDisplay.setTimeZone(TimeZoneUtil.getDefault());
	themeDisplay.setUser(TestPropsValues.getUser());

	_testObjectEntryInfoItemFieldValuesProvider(
		fileEntry, objectAction, objectEntry,
		parentTextObjectFieldNameValue, themeDisplay);
}

private void _testObjectEntryInfoItemFieldValuesProvider(
		FileEntry fileEntry, ObjectAction objectAction,
		ObjectEntry objectEntry, String parentTextObjectFieldNameValue,
		ThemeDisplay themeDisplay)
	throws Exception {

	ServiceContext serviceContext =
		ServiceContextTestUtil.getServiceContext(_group.getGroupId());

	MockHttpServletRequest mockHttpServletRequest =
		new MockHttpServletRequest();

	mockHttpServletRequest.setAttribute(
		WebKeys.THEME_DISPLAY, themeDisplay);

	serviceContext.setRequest(mockHttpServletRequest);

	ServiceContextThreadLocal.pushServiceContext(serviceContext);

	try {
		InfoItemFieldValuesProvider<ObjectEntry>
			infoItemFieldValuesProvider =
				_infoItemServiceRegistry.getFirstInfoItemService(
					InfoItemFieldValuesProvider.class,
					_childObjectDefinition.getClassName());

		InfoItemFieldValues infoItemFieldValues =
			infoItemFieldValuesProvider.getInfoItemFieldValues(objectEntry);

		Assert.assertNotNull(infoItemFieldValues);

		InfoFieldValue<Object> infoLocalizedValue =
			infoItemFieldValues.getInfoFieldValue(objectAction.getName());

		Map<Locale, String> labelMap = objectAction.getLabelMap();

		for (Map.Entry<Locale, String> entry : labelMap.entrySet()) {
			Assert.assertEquals(
				entry.getValue(),
				infoLocalizedValue.getValue(entry.getKey()));
		}

		ObjectField objectField = _objectFieldLocalService.fetchObjectField(
			_childObjectDefinition.getObjectDefinitionId(),
			"attachmentObjectFieldName");

		InfoFieldValue<Object> downloadURLInfoFieldValue =
			infoItemFieldValues.getInfoFieldValue(
				objectField.getObjectFieldId() + "#downloadURL");

		Assert.assertEquals(
			HttpComponentsUtil.removeParameter(
				_dlURLHelper.getDownloadURL(
					fileEntry, fileEntry.getFileVersion(), null,
					StringPool.BLANK),
				"t"),
			HttpComponentsUtil.removeParameter(
				String.valueOf(downloadURLInfoFieldValue.getValue()), "t"));

		_assertInfoFieldValue(
			"#fileName", infoItemFieldValues, objectField,
			fileEntry.getFileName());
		_assertInfoFieldValue(
			"#mimeType", infoItemFieldValues, objectField,
			fileEntry.getMimeType());
		_assertInfoFieldValue(
			"#size", infoItemFieldValues, objectField, fileEntry.getSize());

		InfoFieldValue<Object> parentTextObjectFieldNameInfoFieldValue =
			infoItemFieldValues.getInfoFieldValue(
				"parentTextObjectFieldName");

		Assert.assertEquals(
			parentTextObjectFieldNameValue,
			parentTextObjectFieldNameInfoFieldValue.getValue());

		InfoFieldValue<Object> picklistObjectFieldNameInfoFieldValue =
			infoItemFieldValues.getInfoFieldValue(
				"picklistObjectFieldName");

		for (KeyLocalizedLabelPair keyLocalizedLabelPair :
				(ArrayList<KeyLocalizedLabelPair>)
					picklistObjectFieldNameInfoFieldValue.getValue()) {

			Assert.assertEquals(
				_listTypeEntryKey, keyLocalizedLabelPair.getKey());
		}
	}
	finally {
		ServiceContextThreadLocal.popServiceContext();
	}
}
```

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

**Comment by vngoliveira** [comment]
ci:test:sf

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java [hunk 11]
```diff
@@ -114,13 +112,6 @@ public static void tearDownClass() {
 	public void setUp() throws Exception {
 		_group = GroupTestUtil.addGroup();
 
-		ServiceContext serviceContext =
-			ServiceContextTestUtil.getServiceContext(_group.getGroupId());
-
-		serviceContext.setRequest(_getHttpServletRequest(_group));
-
-		ServiceContextThreadLocal.pushServiceContext(serviceContext);
```
Comment (carolmariaabb) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/info/item/provider/test/ObjectEntryInfoItemFieldValuesProviderTest.java:122]: @vngoliveira as explained, please also delete the `tearDown` method

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

**Comment by vngoliveira** [comment]
ci:test:sf

**Review by carolmariaabb — APPROVED** [review]

**Comment by carolmariaabb** [comment]
ci:test:object

**Comment by carolmariaabb** [comment]
ci:test:relevant

**Comment by carolmariaabb** [comment]
ci:forward:force

**Pull request closed without merging**